### ISSUE TITLE
run XYZ+normal

### ISSUE THREAD

**Issue by wyjkiki** [charlesq34/pointnet2#41]
Hello, can you update the README.md for running XYZ+normal?
Thank you so much!

**Comment by lokalmatador** [comment]
+1 for doing so. I'm also trying to run the net with normals but it crashes in sample_and_group_all. Trying to fix it if possible.

**Comment by programmerkiki** [comment]
I am very happy that someone is doing the same thing. What changes have you made on the basis of the source code? @lokalmatador

**Comment by lokalmatador** [comment]
So far I only used the cls_ssg model to build upon. Looking at the model implementation, especially placeholder inputs, one can see that the channel size is hardcoded as 3 and does not depend on either choosing to use normals or not:

`pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 3))`

Hence, the model can only work with 3 channels; I changed that to reflect the user's choice of using normals or not.

After fixing that, the next error I got was from inside pointnet_utils.py, more specifically in line 78 of `sample_and_group_all` which contained another hardcoded 3; changing that to 6 resolved that issue and now I'm looking at tf_sampling.py (line 56) as the underlying TF operator throws the error that it only can handle 3 channels and not 6. 

All this is done under the assumption that if selecting normals, the channel size should increase to 6. 

Anyways, it'd be great if the developers of pointnet++ could help in this discussion, as in the paper they mention to also use normal information, which the code however does not seem to support properly which is kinda disappointing (yet, being a researcher myself I know how difficult it is to keep everything up to date...).

Cheers!

**Comment by programmerkiki** [comment]
Yes, I have also modified the values you mentioned before, as well as the parameters in the file pointnet_util.py. 

grouped_xyz = tf.reshape(xyz, (batch_size, 1, nsample, 3))

But these changes also did not work. 

It is expected that the author will be able to update the detailed steps into the README.

**Comment by lokalmatador** [comment]
It kind of all boils down to the tf operators they added which - depending on how you change them (3 or 6 channels) partly work or not. Let's hope they find the time.

Have a nice week-end!

**Comment by programmerkiki** [comment]
Look forward to it together.

Have a nice week-end!

**Comment by xuyongzhi** [comment]
Look forward +1

**Comment by qw245** [comment]
I did exactly the same thing.  Hope the authors have some time to revise the code.

**Comment by isakdiaz** [comment]
Has anyone gotten the normals and 6 channels to work on this?

**Comment by ljjyxz123** [comment]
@lokalmatador , I tried to run XYZ+normal dataset,
the open source code here is in default using XYZ only.
And if you want to run the XYZ+normal dataset, 
you may need to modify several positions including the channel issues, for which the author did not deal with it in this open source code.

**Comment by lokalmatador** [comment]
@ljjyxz123 did you succeed in running the net with xyz+normal information? I tried to modify the code, but unfortunately (see above) did not succeed. Similar attempts by other people also seemed to fail.

**Comment by ljjyxz123** [comment]
@lokalmatador , I didn't try to run with xyz+normal datasets.
I am modifying the code to apply to my own dataset now, after that, I will try that. :)

**Comment by lokalmatador** [comment]
@ljjyxz123 thanks for letting me know. In case, we could join our efforts.

**Comment by ljjyxz123** [comment]
@lokalmatador , Hi, what kind of instant messager do you use, maybe we can contact using that tool for much faster communication. :)

**Comment by lokalmatador** [comment]
Skype

**Comment by programmerkiki** [comment]
> @ljjyxz123 thanks for letting me know. In case, we could join our efforts.

Hi,
Do you have any progress about the issue?
Look forward.

**Comment by zmzhouXJTU** [comment]
@ljjyxz123 
Hi，can you post your code that apply your dataset?

**Comment by chunhuaqiushi1989** [comment]
OK,well, maybe we face the same problem, my own data only include the x,y,z,intensity, and I don't know how to use with the author's pointnet?

**Comment by wusuoweima** [comment]
Look forward +1

**Comment by zaiweizhang** [comment]
Look forward +1

**Comment by sheshappanavar** [comment]
The following changes are recommended for the normals to work.

1) models folder change the below lines in respective files

pointnet2_cls_ssg.py: line 16 ==> pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 3)) ==> change from 3 to 6

pointnet2_cls_msg.py: line 12 ==> pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 3)) ==> change 3 to 6

2) in tf_ops/sampling folder

tf_sampling.ccp: line 105 ==> OP_REQUIRES(context,inp_tensor.dims()==3 && inp_tensor.shape().dim_size(2)==3,errors::InvalidArgument("FarthestPointSample expects (batch_size,num_points,3) inp shape")); ==> only change inp_tensor.shape().dim_size(2)==3 to inp_tensor.shape().dim_size(2)==6

tf_sampling.ccp: line 131 ==> OP_REQUIRES(context,inp_tensor.dims()==3 && inp_tensor.shape().dim_size(2)==3,errors::InvalidArgument("GatherPoint expects (batch_size,num_points,3) inp shape")); ==> only change inp_tensor.shape().dim_size(2)==3 to inp_tensor.shape().dim_size(2)==6

3) in tf_ops/grouping folder

tf_grouping.cpp: line 79 ==> OP_REQUIRES(context, xyz1_tensor.dims()==3 && xyz1_tensor.shape().dim_size(2)==3, errors::InvalidArgument("QueryBallPoint expects (batch_size, ndataset, 3) xyz1 shape.")); ==> only change xyz1_tensor.shape().dim_size(2)==3 to xyz1_tensor.shape().dim_size(2)==6

tf_grouping.cpp: line 84 ==> OP_REQUIRES(context, xyz2_tensor.dims()==3 && xyz2_tensor.shape().dim_size(2)==3, errors::InvalidArgument("QueryBallPoint expects (batch_size, npoint, 3) xyz2 shape.")); ==> only change xyz2_tensor.shape().dim_size(2)==3 to xyz2_tensor.shape().dim_size(2)==6

4) in the utils folder

pointnet_util.py: line 75 ==> grouped_xyz = tf.reshape(xyz, (batch_size, 1, nsample, 3)) ==> change from 3 to 6


Note: please compile .cpp file once you modify them to get .so files
P.S: This has worked for me. I hope it works for you all. All the Best.

**Comment by sheshap** [comment]
1) Set FLAGS.normal = True in train files.

2) Replace 
l0_xyz = point_cloud
l0_points = None

by 
l0_xyz = tf.slice(point_cloud, [0,0,0], [-1,-1,3])
l0_points = tf.slice(point_cloud, [0,0,3], [-1,-1,3])

in model files.

3) Replace 
pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 3))

with
pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 6))

in model files

**Comment by lyy-1997-hub** [comment]
> 1. Set FLAGS.normal = True in train files.
> 2. Replace
>    l0_xyz = point_cloud
>    l0_points = None
> 
> by
> l0_xyz = tf.slice(point_cloud, [0,0,0], [-1,-1,3])
> l0_points = tf.slice(point_cloud, [0,0,3], [-1,-1,3])
> 
> in model files.
> 
> 1. Replace
>    pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 3))
> 
> with
> pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 6))
> 
> in model files

thanks！！！！！！！I succeeded！！！！！I love u!!!!

**Comment by Rui-K** [comment]
> 1. Set FLAGS.normal = True in train files.
> 2. Replace
>    l0_xyz = point_cloud
>    l0_points = None
> 
> by l0_xyz = tf.slice(point_cloud, [0,0,0], [-1,-1,3]) l0_points = tf.slice(point_cloud, [0,0,3], [-1,-1,3])
> 
> in model files.
> 
> 3. Replace
>    pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 3))
> 
> with pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 6))
> 
> in model files

This is very useful!! But there is no need to Set FLAGS.normal = True in train files.  Just use --normal and that will make it True.

**Comment by pratibhashinde** [comment]
I have a question: How to use rgb data along with xyz?

Please help me out.

Thanks.

**Comment by jrojas9206** [comment]
> > 1. Set FLAGS.normal = True in train files.
> > 2. Replace
> >    l0_xyz = point_cloud
> >    l0_points = None
> > 
> > by
> > l0_xyz = tf.slice(point_cloud, [0,0,0], [-1,-1,3])
> > l0_points = tf.slice(point_cloud, [0,0,3], [-1,-1,3])
> > in model files.
> > 
> > 1. Replace
> >    pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 3))
> > 
> > with
> > pointclouds_pl = tf.placeholder(tf.float32, shape=(batch_size, num_point, 6))
> > in model files
> 
> thanks！！！！！！！I succeeded！！！！！I love u!!!!

Did you get a good performance after these changes? I  tried but the model never converged :\
